File: android/src/main/kotlin/com/exelbid/flutter/InstanceRegistry.kt
package com.exelbid.flutter

/**
 * Thread-safe registry holding a strong reference to each live full-screen ad
 * handle (mediated/regular interstitial, video) until Dart disposes it. Mirrors
 * iOS `InstanceRegistry`.
 */
object InstanceRegistry {
    private val handles = HashMap<String, Any>()

    @Synchronized
    fun put(id: String, handle: Any) {
        handles[id] = handle
    }

    @Synchronized
    fun remove(id: String) {
        handles.remove(id)
    }
}

File: ios/exelbid_plugin/Sources/exelbid_plugin/Channels/InstanceRegistry.swift
import Foundation

/// Thread-safe registry mapping instance UUID → handle for full-screen ads.
/// Handles store their own SDK objects and channels; the registry only keeps a
/// strong reference so they stay alive until Dart calls `dispose`.
final class InstanceRegistry {
    static let shared = InstanceRegistry()

    private var handles: [String: AnyObject] = [:]
    private let lock = NSLock()

    private init() {}

    func register(_ handle: AnyObject, forId id: String) {
        lock.lock(); defer { lock.unlock() }
        handles[id] = handle
    }

    func handle(forId id: String) -> AnyObject? {
        lock.lock(); defer { lock.unlock() }
        return handles[id]
    }

    @discardableResult
    func remove(id: String) -> AnyObject? {
        lock.lock(); defer { lock.unlock() }
        return handles.removeValue(forKey: id)
    }
}
